Refactor soundwire endpoint parsing

charleskeepax opened the pull request:
These patches redo the parsing of the soundwire endpoints, the new code parses the endpoints into intermediate structs then generates the DAI links from those. This simplifies things and reduces the number of times we need to spin through the snd_soc_acpi_* structs.

charleskeepax committed:
ASoC: Intel: sof-sdw: Add new code for parsing the snd_soc_acpi structs

The current machine driver code has a lot of loops parsing through
the ACPI structs (snd_soc_acpi_link_adr, snd_soc_acpi_adr_device,
snd_soc_acpi_endpoint), this makes it hard to understand exactly
what information is being extracted and for what purposes. As well
as being slightly inefficient, as the same information is looked
up multiple times. There are also some issues with the handling
of multiple endpoints on a single device, only the first of the
snd_soc_acpi_endpoint structures is currently fully processed by the
driver. This means doing things like aggregating the second endpoint
on a device with another device are not currently possible.

Add new parsing code that will count the devices and endpoints, parse
them into an intermediate datastructure, and then use that to create
the DAI links. This patch does not actually utilise the results of
the parsing, items will be moved across in the following patches.

This parsing is based around two new structures which are temporarily
allocated whilst parsing. Firstly, sof_sdw_endpoint, which represents
a specific endpoint for audio on a device and is more or less directly
equivalent to snd_soc_acpi_endpoint. Secondly, sof_sdw_dailink
which represents a DAI link and contains a linked list of one or
more sof_sdw_endpoints. A single trip through the snd_soc_acpi data
structures is used to populate these.

One important point to note here is the use of the num field in
snd_soc_acpi_endpoint to address sof_sdw_dai_info array in the
sof_sdw_codec_info struct. This expects a one to one mapping between
endpoints on a device and dai infos. It would be fine for a specific
system to not specify an endpoint for all of the dai infos available,
but two endpoints mapping to the same dai info would make not sense.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: intel: soc-acpi: Add missing cs42l43 endpoints

Currently the cs42l43 just specifies a single endpoint, as the current
machine driver only looks at the first endpoint specified. Future
refactoring will process all endpoints, as such proper specification
should be added for all the cs42l43 endpoints.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: Intel: sof_sdw: Move generation of DAI links to new parsing

The only part left using the old parsing code is now the generation of
the actual DAI links. Move this generation over to being based on the
new parsing, which allows the removal of the last of the old parsing
code.

The new DAI link generation is a simple matter of creating a new DAI
link for each sof_sdw_dailink struct, and adding a cpu, a codec, and
a mapping for each sof_sdw_endpoint contained in that sof_sdw_dailink.
Note that the CPUs can be inferred as the endpoint list is iterated,
because the endpoints were added into the list sorted by link.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: Intel: sof_sdw: Move ignore_pch_dmic to new parsing

Trivial move of the handling of ignore_pch_dmic over to the new parsing
code.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: Intel: sof_sdw: Move append_dai_type to new parsing

append_dai_type should be set to true in any situation that would
result in multiple DAI links existing on a single SoundWire bus,
because the legacy naming used only the bus number to make things
unique. The current code handles this by looking for codecs with
multiple dai_info structs and looking for buses that include multiple
types of device on them. The first of these assumes that all DAIs on a
given device would be in use. The second, with dissimilar aggregation
now being supported, isn't really an accurate check either since those
devices could be aggregated into a single DAI link.

Move the handling for this flag over to the new parsing code and
simplify things a little by looking directly for SoundWire links
that will contain multiple DAI links to set the flag. These changes
should not cause any change in behaviour for any currently supported
systems.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: intel: sof_sdw: Factor out SoundWire DAI creation

Factor out the creation of the SoundWire DAI links into a helper
function. No functional change.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

charleskeepax committed:
ASoC: Intel: sof_sdw: Move counting and codec_conf to new parsing

Use the output of the new snd_soc_acpi struct parsing for the purposes
of counting the number of SoundWire DAI links and physical devices
attached to the SoundWire. These counts are already returned by the
parser so those can just be used directly. But the population of the
codec_conf structures, is moved from the old parsing code over to the
new. As the two parsers currently co-exist it is better to not have
them both attempt to modify the same structures.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

plbossart commented on sound/soc/intel/common/soc-acpi-intel-tgl-match.c:
```diff
@@ -414,11 +414,32 @@ static const struct snd_soc_acpi_link_adr tgl_712_only[] = {
+		.aggregated = 0,
+		.group_position = 0,
+		.group_id = 0,
+	},
```
it wouldn't hurt to describe what those endpoints might be?

charleskeepax replied:
Yeah, I will put some comments in.

charleskeepax replied:
Also realising I should update MTL as well here will do so.

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1475,8 +1475,164 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
+			*num_ends += adr_link->adr_d[i].num_endpoints;
+	}
+
+	dev_dbg(dev, "Found %d devices with %d endpoints\n", *num_devs, *num_ends);
```
It's not very clear to me what those endpoint counts might mean?

You could have e.g. 2 devices and 5 endpoints with completely different configurations...

charleskeepax replied:
I could drop the debug print its not really vital, I feel like the message would get too large if I try to put some explanation in there, I could maybe put a comment in but that feels weird on a debug print. But to answer the question here, really the counts are used for allocating things internally, the number of devices directly equates to the number of codec_confs we would need, and since the new parsing creates a structure for each endpoint we need to know those. Both of those are rather regardless of the topology of the system.

plbossart replied:
ok

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1475,8 +1475,164 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
+static struct sof_sdw_dailink *find_dailink(struct sof_sdw_dailink *dailinks,
+					    const struct snd_soc_acpi_endpoint *new)
+{
+	bool aggregation = !(sof_sdw_quirk & SOF_SDW_NO_AGGREGATION);
```
I don't fully remember the purpose of this NO_AGGREGATION flag. @bardliao is this not an old thing we could remove?

charleskeepax replied:
I certainly have no feeling for how used this quirk is. I would be more than happy to put a patch in to the start of the series to remove it.

bardliao replied:
@plbossart @charleskeepax non aggregation mode is not supported after https://github.com/thesofproject/linux/pull/4682 is merged. So, we should remove it.

charleskeepax replied:
Cool I will update the chain with a patch to do so.

plbossart reviewed (commented):
quick review, it's rather difficult to figure out what is currently missing and what new functionality will be available when these patches are merged.

charleskeepax commented:
I will try to think about rephrasing some of the commit messages, the one on "[Add new code for parsing the snd_soc_acpi structs](https://github.com/thesofproject/linux/pull/4839/commits/7b08fdf8217ceb868ef85b150308066c74488743)" is probably the one that describes the changes the best, to try to summarise the main new thing this series adds is processing all the endpoints on a device, previously if say the 3rd endpoint on a device was to be aggregated with another device that would have not been processed since only the first endpoint was ever looked at. The series was also rather aimed at tidying up the code and pulling more of the functionality together so its a bit easier to follow and add new bits.

charleskeepax committed:
ASoC: Intel: sof_sdw: Remove no longer supported quirk

The SOF_SDW_NO_AGGREGATION quirk is mostly for debug and no longer works
correctly with the current state of the machine driver. Remove it from
the code and add an error message if someone uses it.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

bardliao commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -50,6 +50,7 @@ enum {
 #define SOF_SSP_PORT(x)		(((x) & GENMASK(5, 0)) << 7)
 #define SOF_SSP_GET_PORT(quirk)	(((quirk) >> 7) & GENMASK(5, 0))
+/* Deprecated and no longer supported by the code */
 #define SOF_SDW_NO_AGGREGATION		BIT(14)
```
How about remove `#define SOF_SDW_NO_AGGREGATION		BIT(14)`?

charleskeepax replied:
Yeah can remove it completely if you want, I was just a little concerned about a quirk being re-used in the future could result in weird behaviour when moving between older and newer kernel versions.

plbossart replied:
It's fine to keep as deprecated, we are not under too much pressure in terms of bits used by quirks. We can always reclaim this later.

plbossart reviewed (approved):
LGTM at a high-level, thanks @charleskeepax
